pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,395 @@
+
+			if te, ok := err.(net.Error); ok && te.Timeout() {
+				_ = c.Close(NoFlush, OnWriteTimeout)
+			}
+			//other write errs not close connection, because readbuffer may have unread data, wait for readloop close connection,
+			return
+		}
+	}
+}
+
+func (c *Connection) appendBuffer(buf *iobufferpool.StreamBuffer) {
+	if buf == nil {
+		return
+	}
+	c.ioBuffers = append(c.ioBuffers, buf)
+	c.writeBuffers = append(c.writeBuffers, buf.Bytes())
+}
+
+// Close connection close function
+func (c *Connection) Close(ccType CloseType, event ConnectionEvent) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Errorf("tcp connection close panic, err: %+v\n%s", r, string(debug.Stack()))
+		}
+	}()
+
+	if ccType == FlushWrite {
+		_ = c.Write(iobufferpool.NewEOFStreamBuffer())
+		return nil
+	}
+
+	if !atomic.CompareAndSwapUint32(&c.closed, 0, 1) {
+		return nil
+	}
+
+	// close tcp rawConn read first
+	logger.Debugf("tcp connection closed(%s), local addr: %s, remote addr: %s",
+		event, c.localAddr.String(), c.remoteAddr.String())
+	if conn, ok := c.rawConn.(*limitlistener.Conn); ok {
+		_ = conn.Conn.(*net.TCPConn).CloseRead() // client connection is wrapped by limitlistener.Conn
+	} else {
+		_ = c.rawConn.(*net.TCPConn).CloseRead()
+	}
+
+	// close rawConn recv, then notify read/write loop to exit
+	close(c.connStopChan)
+	_ = c.rawConn.Close()
+	c.lastBytesSizeRead, c.lastWriteSizeWrite = 0, 0
+
+	if c.onClose != nil {
+		c.onClose(event)
+	}
+	return nil
+}
+
+func (c *Connection) doReadIO() (bufLen int, err error) {
+	if c.readBuffer == nil {
+		c.readBuffer = tcpBufferPool.Get().([]byte)[:iobufferpool.DefaultBufferReadCapacity]
+	}
+
+	// add read deadline setting optimization?
+	// https://github.com/golang/go/issues/15133
+	_ = c.rawConn.SetReadDeadline(time.Now().Add(15 * time.Second))
+	return c.rawConn.(io.Reader).Read(c.readBuffer)
+}
+
+func (c *Connection) doWrite() (int64, error) {
+	_ = c.rawConn.SetWriteDeadline(time.Now().Add(15 * time.Second))
+	bytesSent, err := c.doWriteIO()
+	if err != nil && atomic.LoadUint32(&c.closed) == 1 {
+		return 0, nil
+	}
+
+	if bytesBufSize := c.writeBufLen(); bytesBufSize != c.lastWriteSizeWrite {
+		c.lastWriteSizeWrite = bytesBufSize
+	}
+	return bytesSent, err
+}
+
+func (c *Connection) writeBufLen() (bufLen int) {
+	for _, buf := range c.writeBuffers {
+		bufLen += len(buf)
+	}
+	return
+}
+
+func (c *Connection) doWriteIO() (bytesSent int64, err error) {
+	buffers := c.writeBuffers
+	bytesSent, err = buffers.WriteTo(c.rawConn)
+	if err != nil {
+		return bytesSent, err
+	}
+
+	for i, buf := range c.ioBuffers {
+		c.ioBuffers[i] = nil
+		c.writeBuffers[i] = nil
+		if buf.EOF() {
+			err = iobufferpool.ErrEOF
+		}
+		buf.Release()
+	}
+	c.ioBuffers = c.ioBuffers[:0]
+	c.writeBuffers = c.writeBuffers[:0]
+	return
+}
+
+// ServerConnection wrap tcp connection to backend server
+type ServerConnection struct {
+	Connection
+	connectTimeout time.Duration
+	connectOnce    sync.Once
```

Comment:
why we need to use sync.Once here? do we expect `Connect` to be called multiple times and concurrently?